packages/core/src/modules/drpc/DRPCMessagesApi.ts

```diff
@@ -0,0 +1,173 @@
+import type { DRPCRequest, DRPCRequestObject, DRPCResponse, DRPCResponseObject } from './messages'
+import type { DRPCMessageRecord } from './repository/DRPCMessageRecord'
+import type { ConnectionRecord } from '../connections'
+
+import { AgentContext } from '../../agent'
+import { MessageHandlerRegistry } from '../../agent/MessageHandlerRegistry'
+import { MessageSender } from '../../agent/MessageSender'
+import { OutboundMessageContext } from '../../agent/models'
+import { injectable } from '../../plugins'
+import { isValidDRPCRequestObject } from '../../utils/messageType'
+import { uuid } from '../../utils/uuid'
+import { ConnectionService } from '../connections'
+
+import { DRPCMessageRole } from './DRPCMessageRole'
+import { DRPCMessageHandler } from './handlers'
+import { DRPCErrorCode, DRPCRequestMessage, DRPCResponseMessage } from './messages'
+import { DRPCMessageService } from './services'
+
+@injectable()
+export class DRPCMessagesApi {
+  private drpcMessageService: DRPCMessageService
+  private drpcMethodHandlers: Map<
+    string,
+    (message: DRPCRequestObject) => Promise<DRPCResponseObject | Record<string, never>>
+  > = new Map()
+  private messageSender: MessageSender
+  private connectionService: ConnectionService
+  private agentContext: AgentContext
+
+  public constructor(
+    messageHandlerRegistry: MessageHandlerRegistry,
+    drpcMessageService: DRPCMessageService,
+    messageSender: MessageSender,
+    connectionService: ConnectionService,
+    agentContext: AgentContext
+  ) {
+    this.drpcMessageService = drpcMessageService
+    this.messageSender = messageSender
+    this.connectionService = connectionService
+    this.agentContext = agentContext
+    this.registerMessageHandlers(messageHandlerRegistry)
+
+    // listen for incoming drpc requests and forward them to appropriate handlers
+    this.drpcMessageService.createMessageListener(async ({ message, drpcMessageRecord }) => {
+      if (message instanceof DRPCRequestMessage && drpcMessageRecord.role === DRPCMessageRole.Receiver) {
+        const requests = Array.isArray(message.request) ? message.request : [message.request]
+        const futures = []
+        for (const request of requests) {
+          futures.push(this.handleRequest(request))
+        }
+        const responses = await Promise.all(futures)
+        try {
+          await this.sendDRPCResponse(
+            drpcMessageRecord.connectionId,
+            message.id,
+            responses.length === 1 ? responses[0] : responses
+          )
+        } catch {
+          await this.sendDRPCResponse(drpcMessageRecord.connectionId, message.id, {
```

Comment:
Will it suffice to send a single item in this error response even if there are multiple requests in the same message?